File: stacks_and_queues/stacks/strings/c++/make_string_great/make_string_great.cpp
// make_string_great.cpp
#include "make_string_great.hpp"
#include <string>

char swapCase(char character)
{
    if (std::islower(character))
    {
        return std::toupper(character);
    }
    else
    {
        return std::tolower(character);
    }
}

std::string makeStringGreat(const std::string &s)
{
    if (s.length() == 0)
    {
        return s;
    }

    std::string output = "";
    for (const auto &character : s)
    {
        // If the stack is not empty and the top character form a bad pair with the current character
        if (!output.empty() && swapCase(output.back()) == character)
        {
            // Pop the top character and do not push back
            output.pop_back();
        }
        else
        {
            output.push_back(character);
        }
    }

    return output;
}

File: stacks_and_queues/stacks/strings/python3/make_string_great.py
from typing import List


def make_string_great(s: str) -> str:
    """
    Given a string `s` of lower and upper case English letters. Remove all adjacent characters that one lower case
    letter followed by an upper case letter, or vice versa. 
    
    Parameters
    ----------
    s : str
        A string containing lower case and upper case english characters

    Returns
    -------
    str
        A new string that satisfy the requirements of a good string
    """
    if len(s) == 0:
        return s

    str_stack: List[str] = []
    for char in s:
        # If the stack is not empty and the top character and the current character form a bad pair
        if str_stack and str_stack[-1].swapcase() == char:
            # Pop the top character and do not append the current
            str_stack.pop()
        else:
            str_stack.append(char)

    return "".join(str_stack)
    
def main() -> int:

    test_strings = ["leEeetcode", "abBAcC", "AdG"]
    for s in test_strings:
        new_s = make_string_great(s)
        print(f"The original string: {s}")
        print(f"The new string: {new_s if new_s else 'empty'}")
        print('\n')

    return 0

if __name__ == "__main__":

    main()
